Progress bar (AnsiLoggerWithFooter) display issue

Basically, `AnsiLoggerWithFooter` constantly erases the progress bars (footer) and prints them again.

```
[INFO] Containerizing application to Docker daemon as francium25/test...
[INFO] Executing tasks:
[INFO] [============                  ] 40.0% complete
[INFO] > building image to Docker daemon
```

To do so, it prints out "cursor up" escape sequences to move up the cursor, erases everything to the end of display, and prints an updated footer. This process is repeated. The footer is printed with the `Level.LIFECYCLE` level, for which Maven prefixes `[INFO]`.

While working on #1950, I noticed `eventHandler.dispatch(LogEvent.error(...))` ends up printing `[INFO]` instead of `[ERROR]` when the progress bar is enabled. The reason is related to how `AnsiLoggerWithFooter` erases and re-draws the footer.

Right after `eraseFooter()` in `AnsiLoggerWithFooter.log()`, the display becomes like below momentarily:

```
[INFO] Containerizing application to Docker daemon as francium25/test...
[INFO]
```

Note the last line has the lone `[INFO]` with an empty message. Therefore, the current code moves up the cursor up once to print a log message on that last line.

```java
          // If a previous footer was erased, the message needs to go up a line.
          String messagePrefix = didErase ? CURSOR_UP_SEQUENCE : "";
          messageConsumer.accept(messagePrefix + message);
```

Then, when the log level is not `LIFECYCLE` (e.g., `ERROR`), the display will momentarily look like below:

```
[INFO] Containerizing application to Docker daemon as francium25/test...
[INFO]  THIS LINE SHOULD BE PREFIXED WITH "[ERROR]"
[ERROR] 
```

Now the cursor will be on the last line that has the lone `[ERROR]` prefix. After printing the footer, it immediately becomes

```
[INFO] Containerizing application to Docker daemon as francium25/test...
[INFO]  THIS LINE SHOULD BE PREFIXED WITH "[ERROR]"
[INFO] Executing tasks:
[INFO] [============                  ] 40.0% complete
[INFO] > building image to Docker daemon
```

--- a/jib-core/src/main/java/com/google/cloud/tools/jib/registry/BlobPusher.java
+++ b/jib-core/src/main/java/com/google/cloud/tools/jib/registry/BlobPusher.java
@@ -217,7 +217,7 @@ class BlobPusher {
 
   /**
    * @param location the upload URL
-   * @param blobProgressListener the listener for {@link Blob} push progress
+   * @param writtenByteCountListener the listener for {@link Blob} push progress (written bytes)
    * @return a {@link RegistryEndpointProvider} for writing the BLOB to an upload location
    */
   RegistryEndpointProvider<URL> writer(URL location, Consumer<Long> writtenByteCountListener) {

--- a/jib-plugins-common/src/main/java/com/google/cloud/tools/jib/plugins/common/logging/AnsiLoggerWithFooter.java
+++ b/jib-plugins-common/src/main/java/com/google/cloud/tools/jib/plugins/common/logging/AnsiLoggerWithFooter.java
@@ -37,8 +37,8 @@ class AnsiLoggerWithFooter implements ConsoleLogger {
    */
   private static final int MAX_FOOTER_WIDTH = 50;
 
-  /** ANSI escape sequence for moving the cursor up one line. */
-  private static final String CURSOR_UP_SEQUENCE = "\033[1A";
+  /** ANSI escape sequence template for moving the cursor up multiple lines. */
+  private static final String CURSOR_UP_SEQUENCE_TEMPLATE = "\033[%dA";
 
   /** ANSI escape sequence for erasing to end of display. */
   private static final String ERASE_DISPLAY_BELOW = "\033[0J";
@@ -79,8 +79,9 @@ class AnsiLoggerWithFooter implements ConsoleLogger {
   /**
    * Creates a new {@link AnsiLoggerWithFooter}.
    *
-   * @param messageConsumers map from each {@link Level} to a log message {@link Consumer<String>
-   * @param singleThreadedExecutor a {@link SingleThreadedExecutor} to ensure that all messages are logged in a sequential, deterministic order
+   * @param messageConsumers map from each {@link Level} to a corresponding message logger
+   * @param singleThreadedExecutor a {@link SingleThreadedExecutor} to ensure that all messages are
+   *     logged in a sequential, deterministic order
    */
   AnsiLoggerWithFooter(
       ImmutableMap<Level, Consumer<String>> messageConsumers,
@@ -102,15 +103,15 @@ class AnsiLoggerWithFooter implements ConsoleLogger {
 
     singleThreadedExecutor.execute(
         () -> {
-          boolean didErase = eraseFooter();
-
-          // If a previous footer was erased, the message needs to go up a line.
-          String messagePrefix = didErase ? CURSOR_UP_SEQUENCE : "";
-          messageConsumer.accept(messagePrefix + message);
-
-          for (String footerLine : footerLines) {
-            lifecycleConsumer.accept(BOLD + footerLine + UNBOLD);
+          if (eraseFooter()) {
+            // If a previous footer was erased, the message needs to go up a line. However, we go up
+            // two lines (and then down one line by calling "accept()" once) to correct an issue in
+            // Maven: https://github.com/GoogleContainerTools/jib/issues/1952
+            messageConsumer.accept(String.format(CURSOR_UP_SEQUENCE_TEMPLATE, 2));
           }
+          messageConsumer.accept(message);
+
+          footerLines.forEach(line -> lifecycleConsumer.accept(BOLD + line + UNBOLD));
         });
   }
 
@@ -132,15 +133,14 @@ class AnsiLoggerWithFooter implements ConsoleLogger {
 
     singleThreadedExecutor.execute(
         () -> {
-          boolean didErase = eraseFooter();
-
-          // If a previous footer was erased, the first new footer line needs to go up a line.
-          String newFooterPrefix = didErase ? CURSOR_UP_SEQUENCE : "";
-
-          for (String newFooterLine : truncatedNewFooterLines) {
-            lifecycleConsumer.accept(newFooterPrefix + BOLD + newFooterLine + UNBOLD);
-            newFooterPrefix = "";
+          if (eraseFooter()) {
+            // If a previous footer was erased, the first new footer line needs to go up a line.
+            // However, we go up two lines (and then down one line by calling "accept()" once to
+            // correct an issue in Maven: https://github.com/GoogleContainerTools/jib/issues/1952
+            lifecycleConsumer.accept(String.format(CURSOR_UP_SEQUENCE_TEMPLATE, 2));
           }
+
+          truncatedNewFooterLines.forEach(line -> lifecycleConsumer.accept(BOLD + line + UNBOLD));
 
           footerLines = truncatedNewFooterLines;
         });
@@ -160,12 +160,7 @@ class AnsiLoggerWithFooter implements ConsoleLogger {
     StringBuilder footerEraserBuilder = new StringBuilder();
 
     // Moves the cursor up to the start of the footer.
-    // TODO: Optimize to single init.
-    for (int i = 0; i < footerLines.size(); i++) {
-      // Moves cursor up.
-      footerEraserBuilder.append(CURSOR_UP_SEQUENCE);
-    }
-
+    footerEraserBuilder.append(String.format(CURSOR_UP_SEQUENCE_TEMPLATE, footerLines.size()));
     // Erases everything below cursor.
     footerEraserBuilder.append(ERASE_DISPLAY_BELOW);
 

--- a/jib-plugins-common/src/test/java/com/google/cloud/tools/jib/plugins/common/logging/AnsiLoggerWithFooterTest.java
+++ b/jib-plugins-common/src/test/java/com/google/cloud/tools/jib/plugins/common/logging/AnsiLoggerWithFooterTest.java
@@ -132,21 +132,29 @@ public class AnsiLoggerWithFooterTest {
 
     Assert.assertEquals(
         Arrays.asList(
-            "\033[1mfooter\033[0m",
-            "\033[1A\033[0J",
-            "\033[1Amessage",
-            "\033[1mfooter\033[0m",
-            "\033[1A\033[0J",
-            "\033[1Aanother message",
-            "\033[1mfooter\033[0m"),
+            "\033[1mfooter\033[0m", // single-line footer in bold
+
+            // now triggered by logging a message
+            "\033[1A\033[0J", // cursor up and erase to the end
+            "\033[2A", // up two lines
+            "message",
+            "\033[1mfooter\033[0m", // footer
+
+            // by logging another message
+            "\033[1A\033[0J", // cursor up and erase
+            "\033[2A", // up two
+            "another message",
+            "\033[1mfooter\033[0m"), // footer
         messages);
     Assert.assertEquals(
         Arrays.asList(
             Level.LIFECYCLE,
             Level.LIFECYCLE,
             Level.INFO,
+            Level.INFO,
             Level.LIFECYCLE,
             Level.LIFECYCLE,
+            Level.INFO,
             Level.INFO,
             Level.LIFECYCLE),
         levels);
@@ -163,28 +171,40 @@ public class AnsiLoggerWithFooterTest {
 
     Assert.assertEquals(
         Arrays.asList(
-            "\033[1mfooter\033[0m",
-            "\033[1A\033[0J",
-            "\033[1Amessage",
-            "\033[1mfooter\033[0m",
-            "\033[1A\033[0J",
-            "\033[1A\033[1mtwo line\033[0m",
-            "\033[1mfooter\033[0m",
-            "\033[1A\033[1A\033[0J",
-            "\033[1Aanother message",
-            "\033[1mtwo line\033[0m",
-            "\033[1mfooter\033[0m"),
+            "\033[1mfooter\033[0m", // single-line footer in bold
+
+            // now triggered by logging a warning
+            "\033[1A\033[0J", // cursor up and erase to the end
+            "\033[2A", // up two lines
+            "message",
+            "\033[1mfooter\033[0m", // footer
+
+            // by setting a two-line footer
+            "\033[1A\033[0J", // cursor up and erase
+            "\033[2A", // up two lines
+            "\033[1mtwo line\033[0m", // footer line 1
+            "\033[1mfooter\033[0m", // footer line 2
+
+            // by logging another warning
+            "\033[2A\033[0J", // cursor up twice (to erase two-line footer) and erase
+            "\033[2A", // up two lines
+            "another message",
+            "\033[1mtwo line\033[0m", // footer line 1
+            "\033[1mfooter\033[0m"), // footer line 2
         messages);
     Assert.assertEquals(
         Arrays.asList(
             Level.LIFECYCLE,
             Level.LIFECYCLE,
             Level.WARN,
+            Level.WARN,
             Level.LIFECYCLE,
             Level.LIFECYCLE,
             Level.LIFECYCLE,
             Level.LIFECYCLE,
             Level.LIFECYCLE,
+            Level.LIFECYCLE,
+            Level.WARN,
             Level.WARN,
             Level.LIFECYCLE,
             Level.LIFECYCLE),

